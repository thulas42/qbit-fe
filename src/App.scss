.App {
  max-width: 720px;
  margin: auto;
  &--zar{
    background: #f2f2f2;
    border-radius:5px;
    font-size:4rem;
  }
}

