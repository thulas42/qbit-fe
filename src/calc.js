export const calcZar = (curr, amount, rands, ) => {
    let temp = ( rands / curr) * amount;
    return Number(temp.toFixed(2));
}