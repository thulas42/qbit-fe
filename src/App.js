import React, { useEffect, useState } from 'react';
import * as Yup from 'yup';
import { Formik } from 'formik';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Select from '@material-ui/core/Select';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormControl from '@material-ui/core/FormControl';
import { calcZar } from './calc';

const converstionSchema = Yup.object().shape({
  amount: Yup.number().required('Amount is required'),
  currency: Yup.string().required('Currency is required')
})

const App = () => {

  const [ratesList, setRates] = useState(0);
  const [amount, setAmount] = useState('');
  const [zar, setZar] = useState('');

  const fetchMyApi = async () => {
    try {
      let response = await fetch('http://localhost:4000/currencies');
      let res = await response.json();
      let ratesArr = await Object.keys(res.rates).map((key) => [key, res.rates[key]]);
      let currentZar = ratesArr.find(item => item[0] == 'ZAR')
      setZar(currentZar[1])
      setRates(ratesArr);
    } catch (error) { throw error.message };
  }

  useEffect(() => {
    fetchMyApi();
  }, [])

  return (
    <div className="App">
      <Formik
        initialValues={{
          amount: '',
          currency: ''
        }}
        validationSchema={converstionSchema}
        onSubmit={(values, { setSubmitting }) => {
          setTimeout(() => {
            setSubmitting(false);
            let zarTotal = calcZar(values.currency, parseInt(values.amount), zar)
            setAmount(zarTotal);
          }, 400);
        }}
      >
        {({
          values,
          errors,
          touched,
          handleChange,
          handleSubmit,
        }) => (
          <>
            <div className="App--zar">R{amount ? amount : 0.00}</div>
            <form onSubmit={handleSubmit}>
              <TextField
                fullWidth
                id="amount"
                name="amount"
                label="Enter amount"
                value={values.amount}
                onChange={handleChange}
                error={errors.amount && Boolean(errors.amount)}
                helperText={touched.amount && errors.amount}
              />
              <FormControl error={errors.currency ? true : false} fullWidth>
                <InputLabel
                  style={{ disableAnimation: false }}
                  disableAnimation={false}
                >
                  Select currency
                </InputLabel>                
                <Select
                  fullWidth
                  id="currency"
                  name="currency"
                  label="currency"
                  type="currency"
                  onChange={handleChange}
                  error={Boolean(errors.currency)}
                  helperText={touched.currency && Boolean(errors.currency)}
                >

                  <MenuItem selected disabled value="">
                    <em>None</em>
                  </MenuItem>
                  {ratesList.length > 0 && ratesList.map(key => (
                    <MenuItem key={key} value={key[1]}>{key[0]}</MenuItem>))}
                </Select>
                <FormHelperText>{touched.currency && errors.currency}</FormHelperText>
              </FormControl>
              <Button color="primary" variant="contained" type="submit">
                Convert
              </Button>
            </form>
          </>

        )}
      </Formik>

    </div>
  );
};

export default App