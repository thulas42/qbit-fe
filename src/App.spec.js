import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import {calcZar}from './calc';

describe('App', () => {
    test('snapshot renders', () => {
      const component = renderer.create(<App />);
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('Conversion permutations test',  ()=>{
        const totalAmount =  calcZar(1,1,1.18074,1)
        expect(totalAmount).toEqual(1.18);
    })

    test('Decimal places test', ()=>{
        const totalAmount =  calcZar(1,1,1.18074,1)
        const regex = new RegExp("^\s*-?\d+(\.\d{1,2})?\s*$")
        expect((regex).test(totalAmount));
    })
  });