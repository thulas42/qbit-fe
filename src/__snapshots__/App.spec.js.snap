// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div
  className="App"
>
  <div
    className="App--zar"
  >
    R
    0
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
        htmlFor="amount"
        id="amount-label"
      >
        Enter amount
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          id="amount"
          name="amount"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
        style={
          Object {
            "disableAnimation": false,
          }
        }
      >
        Select currency
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="currency"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="currency"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          name="currency"
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiFormHelperText-root"
      />
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Convert
      </span>
    </button>
  </form>
</div>
`;
